### ISSUE TITLE
Breakpad Use Case

### ISSUE THREAD

**Issue by Jake-Shadle** [Microsoft/microsoft-pdb#2]
The README noted you would like issues pertaining to use cases, so here is a slightly complicated one. :)

If you are not aware, [Breakpad symbol files](https://chromium.googlesource.com/breakpad/breakpad/+/master/docs/symbol_files.md) are a way to get cross platform support for decoding callstacks from crash dumps/bug reports etc. The Breakpad symbol files are also ~20% of the size of the PDB they are generated from since they only contain a subset of the information.

I helped create a PDB reader [here](https://github.com/luser/dump_syms) that has most of the support needed, but there are some things missing, as well as edge cases that sometimes crop up due to slight differences between eg. the standard MSVC++ Windows compiler and the XBox 1 compiler. (Though this will hopefully be resolved once the XB1 uses the same exact compiler which is coming soon in VS2015).

The reason this was made was because of several reasons:

* Abysmal performance of DIA.
  * All strings are heap allocated, and they are all considered wide strings, even though I have yet to find string records for eg. function names/type names etc that are anything but standard ASCII.
  * The COM interface is an impediment to linear processing, it was only designed for doing single lookups of individual pieces of information it seems.
* Ability to process anywhere
  * DIA makes the symbol processing have a hard dependency on Windows, most shops would like the option of being able to do the Breakpad generation on a Linux box if desired.

So basically I guess I am asking for...an almost full spec of the format. :) All of the information I used to reverse engineer parts of the format was either old, or only showed some of the pertinent data (eg. [MS CCI](http://ccimetadata.codeplex.com/SourceControl/latest#Sources/PdbReader/PdbReader.cs).

**Comment by luser** [comment]
This is a great start! Some things that would be very useful but aren't currently present include:
* The DBI header (+ DBI debug header)
* The format of the streams referenced from those  headers (GSI, PSI, symbol stream, etc)

**Comment by AndrewPardoe** [comment]
Thanks for the suggestions, Jack-Shadle and luser. 

Jack-Shadle, an almost-full spec of the format is probably not going to come out in the near future : ) Are there specific items you'd like more information on? Does luser's list cover your needs?

**Comment by Jake-Shadle** [comment]
Yes, those probably cover the most important bits. There are also cases where data in some streams are back to back, and in other streams are padded to be aligned on eg. 4 byte boundaries, so any clarification on what the actual expectations are for different streams would help a lot since it would allow for better error detection rather than just "we think the data might be wrong, but we aren't sure because we wrote our own parser" :smile:

**Comment by skochinsky** [comment]
Some third-partty documentation is available [here](http://code.google.com/p/pdbparser/w/list)

**Comment by AndrewPardoe** [comment]
Interesting, @skochinsky. I see this is BSD-3 licensed. I'll see if we can build from it or rehost it.

**Comment by skochinsky** [comment]
[This overview](http://lists.llvm.org/pipermail/cfe-dev/2015-October/045780.html) should probably be added too. And maybe [the original (16-bit) CodeView spec](ftp://ftp.openwatcom.org/pub/devel/docs/CodeView.pdf). And "Visual C++ 5.0 Symbolic Debug Information Specification" from MSDN 2001.

**Comment by assarbad** [comment]
Hope this is not out of line, but has anybody also considered [](http://undocumented.rawol.com/)? In particular the PDF version from there? The link to the download seems to be dead, but I'm sure I can dig up a copy which I downloaded some years ago.

**Comment by AndrewPardoe** [comment]
@assarbad if you find a live link to the PDF I'll happily add it into the external resources doc. (Or you could just PR with the link.)

**Comment by assarbad** [comment]
@AndrewPardoe: hmm, that's the problem. I do have the files still (separate PDFs) and I also created a combined PDF which includes the two downloads. But there doesn't seem to be a live version of it around anymore.

However, this [download](http://undocumented.rawol.com/win_pdbx.zip) is likely useful as documentation - and it's still live.

**Comment by moyix** [comment]
The site seems somewhat dubious, but the PDF is mirrored here:

http://index-of.es/eBooks/Undocumented.Windows.2000.Secrets.-.The.Programmers.Cookbook.pdf

**Comment by assarbad** [comment]
@moyix That's not the PDF, btw. The stuff that got released by the author himself were thirteen separate PDF files. The one you are linking is a *scanned* copy. The one by the author even included the cut marks in the index (the other PDF pages were already cropped).

[This ZIP](https://bitbucket.org/assarbad/scripts/downloads/Undocumented%20Windows%202000%20Secrets%20by%20Sven%20B.%20Schreiber%20(undocumented.rawol.com).zip) contains the stuff I downloaded in late 2009 from the author's website.
